src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -536,30 +536,38 @@ object Parsers {
      */
     def path(thisOK: Boolean, finish: Tree => Tree = id): Tree = {
       val start = in.offset
-      def handleThis(name: TypeName) = {
+      var tTpeName: TypeName = tpnme.EMPTY
+
+      def acceptThis(): Tree = {
         in.nextToken()
-        val t = atPos(start) { This(name) }
-        if (!thisOK && in.token != DOT) syntaxError("`.' expected")
-        manyDotSelectors(t, finish)
+        atPos(start) { This(tTpeName) }
       }
-      def handleSuper(name: TypeName) = {
+      def acceptSuper(): Tree = {
         in.nextToken()
         val mix = mixinQualifierOpt()
-        val t = atPos(start) { Super(This(name), mix) }
-        accept(DOT)
-        manyDotSelectors(selector(t), finish)
+        atPos(start) { Super(This(tTpeName), mix) }
       }
-      if (in.token == THIS) handleThis(tpnme.EMPTY)
-      else if (in.token == SUPER) handleSuper(tpnme.EMPTY)
-      else {
+      def thisSuperOrElse(alternative: => Tree): Tree =
+        if (in.token == THIS && !thisOK)
+          someDotSelectors(acceptThis(), finish)
+        else if (in.token == THIS)
+          manyDotSelectors(acceptThis(), finish)
```

Comment:
Aaah sorry, I'm confused and retract, I've just read that:

> In particular, the same syntax error would eventually be raised by accept(DOT) in someDotSelectors.

I've also read the ScalaDoc for `thisOK`.

And now I've figured what is happening. The distinction between `someDotSelectors` and `manyDotSelectors` is tricky.